{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "File:    hw5_part1.py\n",
    "Author:  gia Bao VO\n",
    "Date:    10/12/20\n",
    "Section: 31\n",
    "E-mail:  giabaov1@umbc.edu\n",
    "Description:\n",
    "  This program will \n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list1 =[]\n",
    "    for i in range(1,len(the_sequence),CODON_LENGTH):\n",
    "        result = the_sequence[i:CODON_LENGTH]\n",
    "        if(result in START):\n",
    "            j = i + CODON_LENGTH\n",
    "            list1.append(result)\n",
    "            end_codon = the_sequence[j:CODON_LENGTH]\n",
    "            while(end_codon not in STOP):\n",
    "                end_codon = the_sequence[j:CODON_LENGTH]\n",
    "                result = result + end_codon\n",
    "                j = j + CODON_LENGTH\n",
    "            list1.append(result)\n",
    "            print(list1)\n",
    "    return list1\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if i in len(codons):\n",
    "    for [i] in i:\n",
    "        if (randInt.rand == 1):\n",
    "            genes.append(\"A\")\n",
    "        elif (randInt.rand == 2):\n",
    "            genes.append(\"T\")\n",
    "    \n",
    "        elif (randInt.rand == 3):\n",
    "            genes.append(\"C\")\n",
    "        elif (randInt.rand == 4):\n",
    "            genes.append(\"G\")\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "How many codons do you want to create? (or stop to end, seq to enter your own sequence)4\n",
      "CCAGAATTGAGA\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "from random import choice, seed\n",
    "\n",
    "\n",
    "if len(sys.argv) >= 2:\n",
    "    seed(sys.argv[1])\n",
    "\n",
    "# USE IF YOU ARE TESTING AND DON'T WANT TO USE COMMAND LINE ARGUMENTS\n",
    "# seed(input('What seed do you want to use? '))\n",
    "# END SECTION\n",
    "\n",
    "STOP_PARAM = 'stop'\n",
    "SEQ_PARAM = 'seq'\n",
    "NUCLEOTIDES = ['A', 'T', 'C', 'G']\n",
    "# your constants should start here\n",
    "START = ['ATG']\n",
    "STOP = ['TAA', 'TGA', 'TAG']\n",
    "CODON_LENGTH = 3\n",
    "\n",
    "\n",
    "def extract_genes(sequence):\n",
    "    extraction_list =[]\n",
    "    for x in range(0,len(sequence),CODON_LENGTH):\n",
    "        result = sequence[x:x + CODON_LENGTH]\n",
    "        if result in START:\n",
    "            y = x + CODON_LENGTH\n",
    "            extraction_list.append(result)\n",
    "            end_codon = sequence[y:CODON_LENGTH]\n",
    "            x=0\n",
    "            while(end_codon not in STOP):\n",
    "                end_codon = sequence[y:CODON_LENGTH]\n",
    "                result = result + end_codon\n",
    "                y = x + CODON_LENGTH\n",
    "                extraction_list.append(result)\n",
    "            print(extraction_list)\n",
    "    return extraction_list       \n",
    "    pass\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    length_or_stop = input('How many codons do you want to create? (or stop to end, seq to enter your own sequence)')\n",
    "    while length_or_stop.lower() != STOP_PARAM:\n",
    "        try:\n",
    "            if length_or_stop.lower() == SEQ_PARAM:\n",
    "                the_sequence = input('Enter your own sequence: ').upper()\n",
    "                if len(the_sequence) % 3 != 0:\n",
    "                    raise ValueError('The length of the string must be divisible by 3')\n",
    "                if any(x not in NUCLEOTIDES for x in the_sequence):\n",
    "                    raise ValueError('The sequence must contain only A, T, C, G')\n",
    "            else:\n",
    "                the_sequence = ''.join([choice(NUCLEOTIDES) for _ in range(3 * int(length_or_stop))])\n",
    "            print(the_sequence)\n",
    "            print(extract_genes(the_sequence))\n",
    "        except ValueError:\n",
    "            print('You entered a non-STOP non-integer, try again. ')\n",
    "        length_or_stop = input('How many codons do you want to create? (or stop to end, seq to enter your own sequence)')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "How many codons do you want to create? (or stop to end, seq to enter your own sequence)4\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'STOP_PARAM' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-6-8763bd343cc2>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     67\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'__main__'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     68\u001b[0m     \u001b[0mlength_or_stop\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0minput\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'How many codons do you want to create? (or stop to end, seq to enter your own sequence)'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 69\u001b[1;33m     \u001b[1;32mwhile\u001b[0m \u001b[0mlength_or_stop\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlower\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[0mSTOP_PARAM\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     70\u001b[0m         \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     71\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mlength_or_stop\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlower\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mSEQ_PARAM\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'STOP_PARAM' is not defined"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "File:    hw5_part1.py\n",
    "Author:  gia Bao VO\n",
    "Date:    10/12/20\n",
    "Section: 31\n",
    "E-mail:  giabaov1@umbc.edu\n",
    "Description:\n",
    "  This program will divide sequences and loop it and make genetic attraction.\n",
    "\"\"\"\n",
    "def extract_genes(sequence):\n",
    "    \"\"\"\n",
    "    This function should return a list of genes that start with the\n",
    "    start codon and end with one of the three stop codons.\n",
    "    :param sequence: a string of nucleotides\n",
    "    :return: a list of \"genes\"\n",
    "    \"\"\"\n",
    "    # holds the divided nucleotides sequence\n",
    "    sequence_list = []  #nucleo_trio\n",
    "    \n",
    "    # holds the combined strings\n",
    "    new_sequence = []   #new_starting_position\n",
    "    \n",
    "    # breaks the sequence up into 3s and appends into the a list (nucleo_trio_list)\n",
    "    x = 0\n",
    "    beginning_of_codon = 0 \n",
    "    ending_of_codon = 3\n",
    "    next_sequence = 3\n",
    "    while x < len(the_sequence):\n",
    "        sequence_list.append(the_sequence[beginning_of_codon:ending_of_codon])\n",
    "        beginning_of_codon += next_sequence\n",
    "        ending_of_codon += next_sequence \n",
    "        x += next_sequence \n",
    "\n",
    "    # combines the 3 codons together into a one joined string\n",
    "    final_sequence= ''\n",
    "    \n",
    "    x = 0\n",
    "    for y in range(len(sequence_list)):\n",
    "        if sequence_list[y] in START:\n",
    "\n",
    "            # remembers where the START codon begins in the list and stores as int value\n",
    "            x = y\n",
    "\n",
    "            while x < len(sequence_list):\n",
    "                \n",
    "                #runs until there is a stop codon\n",
    "                if sequence_list[x] not in STOP:\n",
    "\n",
    "                    final_sequence += sequence_list[x]\n",
    "                    x += 1\n",
    "                    \n",
    "                else:\n",
    "\n",
    "                    final_sequence += sequence_list[x]\n",
    "\n",
    "                    # jumps out of the while loop\n",
    "                    x = len(sequence_list)\n",
    "\n",
    "                    # adds the combined sequence into a list to save\n",
    "                    new_sequence.append(final_sequence) \n",
    "\n",
    "        # resets the combine string\n",
    "        final_sequence=''\n",
    "        \n",
    "    return(new_sequence)   \n",
    "    pass \n",
    "if __name__ == '__main__':\n",
    "    length_or_stop = input('How many codons do you want to create? (or stop to end, seq to enter your own sequence)')\n",
    "    while length_or_stop.lower() != STOP_PARAM:\n",
    "        try:\n",
    "            if length_or_stop.lower() == SEQ_PARAM:\n",
    "                the_sequence = input('Enter your own sequence: ').upper()\n",
    "                if len(the_sequence) % 3 != 0:\n",
    "                    raise ValueError('The length of the string must be divisible by 3')\n",
    "                if any(x not in NUCLEOTIDES for x in the_sequence):\n",
    "                    raise ValueError('The sequence must contain only A, T, C, G')\n",
    "            else:\n",
    "                the_sequence = ''.join([choice(NUCLEOTIDES) for _ in range(3 * int(length_or_stop))])\n",
    "            print(the_sequence)\n",
    "            print(extract_genes(the_sequence))\n",
    "        except ValueError:\n",
    "            print('You entered a non-STOP non-integer, try again. ')\n",
    "        length_or_stop = input('How many codons do you want to create? (or stop to end, seq to enter your own sequence)')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "How many codons do you want to create? (or stop to end, seq to enter your own sequence)stop\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'STOP_PARAM' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-7-d928438c15a3>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     59\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'__main__'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     60\u001b[0m     \u001b[0mlength_or_stop\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0minput\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'How many codons do you want to create? (or stop to end, seq to enter your own sequence)'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 61\u001b[1;33m     \u001b[1;32mwhile\u001b[0m \u001b[0mlength_or_stop\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlower\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[0mSTOP_PARAM\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     62\u001b[0m         \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     63\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mlength_or_stop\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlower\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mSEQ_PARAM\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'STOP_PARAM' is not defined"
     ]
    }
   ],
   "source": [
    "def extract_genes(sequence):\n",
    "    \"\"\"\n",
    "    This function should return a list of genes that start with the\n",
    "    start codon and end with one of the three stop codons.\n",
    "    :param sequence: a string of nucleotides\n",
    "    :return: a list of \"genes\"\n",
    "    \"\"\"\n",
    "    # holds the divided nucleotides sequence\n",
    "    nucleo_trio_list = []\n",
    "    \n",
    "    # holds the combined strings\n",
    "    new_starting_position = []\n",
    "    \n",
    "    # breaks the sequence up into 3s and appends into the a list (nucleo_trio_list)\n",
    "    i = 0\n",
    "    beginning_of_codon = 0 \n",
    "    ending_of_codon = 3\n",
    "    next_sequence = 3\n",
    "    while i < len(the_sequence):\n",
    "        nucleo_trio_list.append(the_sequence[beginning_of_codon:ending_of_codon])\n",
    "        beginning_of_codon += next_sequence\n",
    "        ending_of_codon += next_sequence \n",
    "        i += next_sequence \n",
    "\n",
    "    # combines the 3 codons together into a one joined string\n",
    "    new_sequence= ''\n",
    "    \n",
    "    i = 0\n",
    "    for j in range(len(nucleo_trio_list)):\n",
    "        if nucleo_trio_list[j] in START:\n",
    "\n",
    "            # remembers where the START codon begins in the list and stores as int value\n",
    "            i = j\n",
    "\n",
    "            while i < len(nucleo_trio_list):\n",
    "                \n",
    "                #runs until there is a stop codon\n",
    "                if nucleo_trio_list[i] not in STOP:\n",
    "\n",
    "                    new_sequence += nucleo_trio_list[i]\n",
    "                    i += 1\n",
    "                    \n",
    "                else:\n",
    "\n",
    "                    new_sequence += nucleo_trio_list[i]\n",
    "\n",
    "                    # jumps out of the while loop\n",
    "                    i = len(nucleo_trio_list)\n",
    "\n",
    "                    # adds the combined sequence into a list to save\n",
    "                    new_starting_position.append(new_sequence) \n",
    "\n",
    "        # resets the combine string\n",
    "        new_sequence=''\n",
    "        \n",
    "    return(new_starting_position)   \n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    length_or_stop = input('How many codons do you want to create? (or stop to end, seq to enter your own sequence)')\n",
    "    while length_or_stop.lower() != STOP_PARAM:\n",
    "        try:\n",
    "            if length_or_stop.lower() == SEQ_PARAM:\n",
    "                the_sequence = input('Enter your own sequence: ').upper()\n",
    "                if len(the_sequence) % 3 != 0:\n",
    "                    raise ValueError('The length of the string must be divisible by 3')\n",
    "                if any(x not in NUCLEOTIDES for x in the_sequence):\n",
    "                    raise ValueError('The sequence must contain only A, T, C, G')\n",
    "            else:\n",
    "                the_sequence = ''.join([choice(NUCLEOTIDES) for _ in range(3 * int(length_or_stop))])\n",
    "            print(the_sequence)\n",
    "            print(extract_genes(the_sequence))\n",
    "        except ValueError:\n",
    "            print('You entered a non-STOP non-integer, try again. ')\n",
    "        length_or_stop = input('How many codons do you want to create? (or stop to end, seq to enter your own sequence)')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
